subsys/usb/device_next/class/loopback.c

```diff
@@ -43,10 +46,133 @@ struct loopback_desc {
+	const struct usb_desc_header **const hs_desc;
+	atomic_t state;
+};
+
+static void lb_update(struct usbd_class_node *c_nd,
+		      uint8_t iface, uint8_t alternate)
+{
+	LOG_DBG("Instance %p, interface %u alternate %u changed",
+		c_nd, iface, alternate);
+}
+
+static int lb_control_to_host(struct usbd_class_node *c_nd,
+			      const struct usb_setup_packet *const setup,
+			      struct net_buf *const buf)
+{
+	if (setup->RequestType.recipient != USB_REQTYPE_RECIPIENT_DEVICE) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (setup->bRequest == LB_VENDOR_REQ_IN) {
+		net_buf_add_mem(buf, lb_buf,
+				MIN(sizeof(lb_buf), setup->wLength));
+
+		LOG_WRN("Device-to-Host, wLength %u | %zu", setup->wLength,
+			MIN(sizeof(lb_buf), setup->wLength));
+
+		return 0;
+	}
+
+	LOG_ERR("Class request 0x%x not supported", setup->bRequest);
+	errno = -ENOTSUP;
+
+	return 0;
+}
+
+static int lb_control_to_dev(struct usbd_class_node *c_nd,
+			     const struct usb_setup_packet *const setup,
+			     const struct net_buf *const buf)
+{
+	if (setup->RequestType.recipient != USB_REQTYPE_RECIPIENT_DEVICE) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (setup->bRequest == LB_VENDOR_REQ_OUT) {
+		LOG_WRN("Host-to-Device, wLength %u | %zu", setup->wLength,
+			MIN(sizeof(lb_buf), buf->len));
+		memcpy(lb_buf, buf->data, MIN(sizeof(lb_buf), buf->len));
+		return 0;
+	}
+
+	LOG_ERR("Class request 0x%x not supported", setup->bRequest);
+	errno = -ENOTSUP;
+
+	return 0;
+}
+
+static int lb_request_handler(struct usbd_class_node *c_nd,
+			      struct net_buf *buf, int err)
+{
+	struct udc_buf_info *bi = NULL;
+
+	bi = (struct udc_buf_info *)net_buf_user_data(buf);
+	LOG_DBG("%p -> ep 0x%02x, len %u, err %d", c_nd, bi->ep, buf->len, err);
+	usbd_ep_buf_free(c_nd->data->uds_ctx, buf);
+
+	return 0;
+}
+
+static void *lb_get_desc(struct usbd_class_node *const c_nd,
+			 const enum usbd_speed speed)
+{
+	struct lb_data *data = usbd_class_get_private(c_nd);
+
+	if (speed == USBD_SPEED_HS) {
+		return data->hs_desc;
+	}
+
+	return data->fs_desc;
+}
+
+static int lb_init(struct usbd_class_node *c_nd)
+{
+	struct usbd_contex *uds_ctx = usbd_class_get_ctx(c_nd);
+	struct lb_data *data = usbd_class_get_private(c_nd);
+	struct loopback_desc *desc = data->desc;
+
+	desc->iad.bFirstInterface = desc->if0.bInterfaceNumber;
+	if (usbd_caps_speed(uds_ctx) == USBD_SPEED_HS) {
+		LOG_INF("FS endpoint descriptor needs to be updated");
+		desc->if0_in_ep.bEndpointAddress = desc->if0_hs_in_ep.bEndpointAddress;
+		desc->if0_out_ep.bEndpointAddress = desc->if0_hs_out_ep.bEndpointAddress;
+	}
+
+	LOG_DBG("Init class instance %p", c_nd);
+
+	return 0;
+}
+
+struct usbd_class_api lb_api = {
+	.update = lb_update,
+	.control_to_host = lb_control_to_host,
+	.control_to_dev = lb_control_to_dev,
+	.request = lb_request_handler,
+	.get_desc = lb_get_desc,
+	.init = lb_init,
+};
 
 #define DEFINE_LOOPBACK_DESCRIPTOR(x, _)					\
 static struct loopback_desc lb_desc_##x = {					\
```

Comment:
After a second thought, it can be kept as-is now, but in the future separated to allow the descriptors that are constant to reside only in code memory and not in RAM (which means both code and RAM, because initialization data must be in code).